File: platforms/android/java/app/msime/client/core/EmojiCatalogModel.java
package app.msime.client;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;

/** Host-only paging and recent-selection policy for the Engine-owned emoji catalog. */
public final class EmojiCatalogModel {
    public static final int COLUMNS = 8;
    public static final int PAGE_SIZE = 64;
    public static final int RECENTS_LIMIT = 24;
    public static final int MAX_TEXT_CODE_POINTS = 32;
    public static final int MAX_ANNOTATION_CODE_POINTS = 1_024;

    public record Category(String group, String title) {}
    public record Item(String text, String annotation, String group) {
        public Item {
            if (text == null || text.isEmpty()
                    || text.codePointCount(0, text.length()) > MAX_TEXT_CODE_POINTS)
                throw new IllegalArgumentException("Invalid emoji catalog text");
            if (annotation == null || annotation.codePointCount(0, annotation.length())
                    > MAX_ANNOTATION_CODE_POINTS)
                throw new IllegalArgumentException("Invalid emoji annotation");
            if (group == null || group.isEmpty() || group.length() > 128)
                throw new IllegalArgumentException("Invalid emoji group");
        }
    }
    public record Page(List<Item> items, int nextOffset, boolean complete) {
        public Page {
            items = List.copyOf(items);
        }
    }

    // Unicode group order; database row sort order interleaves Symbols and Flags.
    private static final List<Category> CATEGORIES = List.of(
        new Category("Smileys and emotion", "笑脸"),
        new Category("People and body", "人物"),
        new Category("Animals and nature", "动物"),
        new Category("Food and drink", "食物"),
        new Category("Travel and places", "旅行"),
        new Category("Activities", "活动"),
        new Category("Objects", "物品"),
        new Category("Symbols", "符号"),
        new Category("Flags", "旗帜")
    );

    private EmojiCatalogModel() {}

    public static List<Category> categories() { return CATEGORIES; }

    public static Page validatePage(
            List<Item> items, int requestedOffset, int limit, long nextOffset, boolean complete) {
        if (requestedOffset < 0 || limit < 1 || limit > 255 || items == null
                || items.size() > limit || nextOffset < requestedOffset
                || nextOffset > (long) requestedOffset + limit
                || nextOffset > Integer.MAX_VALUE || (!complete && nextOffset == requestedOffset))
            throw new IllegalArgumentException("Invalid emoji catalog page");
        return new Page(items, (int) nextOffset, complete);
    }

    /** Most-recent first, deduplicated, and bounded without retaining invalid persisted values. */
    public static List<String> normalizeRecents(List<String> stored) {
        LinkedHashSet<String> unique = new LinkedHashSet<>();
        if (stored != null) {
            for (String text : stored) {
                if (text == null || text.isEmpty()
                        || text.codePointCount(0, text.length()) > MAX_TEXT_CODE_POINTS) continue;
                unique.add(text);
                if (unique.size() == RECENTS_LIMIT) break;
            }
        }
        return List.copyOf(unique);
    }

    public static List<String> recordRecent(List<String> stored, String selected) {
        if (selected == null || selected.isEmpty()
                || selected.codePointCount(0, selected.length()) > MAX_TEXT_CODE_POINTS)
            throw new IllegalArgumentException("Invalid recent emoji");
        ArrayList<String> reordered = new ArrayList<>();
        reordered.add(selected);
        for (String text : normalizeRecents(stored)) {
            if (!selected.equals(text)) reordered.add(text);
            if (reordered.size() == RECENTS_LIMIT) break;
        }
        return List.copyOf(reordered);
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/emoji/EmojiCatalogModel.ts
import { utf8Length } from '../Utf8';

/**
 * Paging and recent-selection policy for the Engine-owned emoji catalog, ported from
 * platforms/android/java/app/msime/client/EmojiCatalogModel.java.
 *
 * Lengths are counted in code points, not UTF-16 units: an emoji is routinely several units long and
 * a sequence with a skin tone or a zero-width joiner is longer still.
 */
export const EMOJI_COLUMNS: number = 8;
export const EMOJI_PAGE_SIZE: number = 64;
export const EMOJI_RECENTS_LIMIT: number = 24;
export const EMOJI_RECENTS_MAX_BYTES: number = 16 * 1024;
export const MAX_TEXT_CODE_POINTS: number = 32;
export const MAX_ANNOTATION_CODE_POINTS: number = 1024;

export interface EmojiCategory {
  readonly group: string;
  readonly title: string;
}

export interface EmojiItem {
  readonly text: string;
  readonly annotation: string;
  readonly group: string;
}

export interface EmojiPage {
  readonly items: EmojiItem[];
  readonly nextOffset: number;
  readonly complete: boolean;
}

function codePointCount(text: string): number {
  let count: number = 0;
  for (const _character of text) {
    count++;
  }
  return count;
}

// Unicode group order; the database row sort order interleaves Symbols and Flags.
const CATEGORIES: EmojiCategory[] = [
  { group: 'Smileys and emotion', title: '笑脸' },
  { group: 'People and body', title: '人物' },
  { group: 'Animals and nature', title: '动物' },
  { group: 'Food and drink', title: '食物' },
  { group: 'Travel and places', title: '旅行' },
  { group: 'Activities', title: '活动' },
  { group: 'Objects', title: '物品' },
  { group: 'Symbols', title: '符号' },
  { group: 'Flags', title: '旗帜' }
];

export class EmojiCatalogModel {
  static categories(): EmojiCategory[] {
    return CATEGORIES;
  }

  static item(text: string, annotation: string, group: string): EmojiItem {
    if (text.length === 0 || codePointCount(text) > MAX_TEXT_CODE_POINTS) {
      throw new Error('Invalid emoji catalog text');
    }
    if (codePointCount(annotation) > MAX_ANNOTATION_CODE_POINTS) {
      throw new Error('Invalid emoji annotation');
    }
    if (group.length === 0 || group.length > 128) {
      throw new Error('Invalid emoji group');
    }
    return { text: text, annotation: annotation, group: group };
  }

  /**
   * A page the Engine returned is checked before it is trusted: an offset that goes backwards, or a
   * page that claims to be incomplete while advancing nowhere, would loop the caller forever.
   */
  static validatePage(items: EmojiItem[], requestedOffset: number, limit: number,
                      nextOffset: number, complete: boolean): EmojiPage {
    if (requestedOffset < 0 || limit < 1 || limit > 255 || items.length > limit
        || nextOffset < requestedOffset || nextOffset > requestedOffset + limit
        || nextOffset > 2147483647 || (!complete && nextOffset === requestedOffset)) {
      throw new Error('Invalid emoji catalog page');
    }
    return { items: items.slice(), nextOffset: nextOffset, complete: complete };
  }

  /** Most-recent first, deduplicated, and bounded without retaining invalid persisted values. */
  static normalizeRecents(stored: string[] | null): string[] {
    const unique: string[] = [];
    if (stored !== null) {
      for (const text of stored) {
        if (text.length === 0 || codePointCount(text) > MAX_TEXT_CODE_POINTS
            || unique.includes(text)) {
          continue;
        }
        unique.push(text);
        if (unique.length === EMOJI_RECENTS_LIMIT) {
          break;
        }
      }
    }
    return unique;
  }

  /**
   * Reads the small app-private recents document without trusting its shape.
   *
   * Recents are convenience state rather than a reason to make the keyboard fail: a malformed or
   * over-sized file is treated as empty, while every value that survives still goes through the
   * same code-point and duplicate bounds as values selected in this process.
   */
  static parseRecents(document: string | null): string[] {
    if (document === null || document.length === 0
        || utf8Length(document) > EMOJI_RECENTS_MAX_BYTES) {
      return [];
    }
    try {
      const decoded: unknown = JSON.parse(document);
      if (!Array.isArray(decoded)) {
        return [];
      }
      const values: string[] = [];
      for (const value of decoded) {
        if (typeof value === 'string') {
          values.push(value);
        }
      }
      return EmojiCatalogModel.normalizeRecents(values);
    } catch {
      return [];
    }
  }

  static serializeRecents(stored: string[] | null): string {
    return JSON.stringify(EmojiCatalogModel.normalizeRecents(stored));
  }

  static recordRecent(stored: string[] | null, selected: string): string[] {
    if (selected.length === 0 || codePointCount(selected) > MAX_TEXT_CODE_POINTS) {
      throw new Error('Invalid recent emoji');
    }
    const reordered: string[] = [selected];
    for (const text of EmojiCatalogModel.normalizeRecents(stored)) {
      if (selected !== text) {
        reordered.push(text);
      }
      if (reordered.length === EMOJI_RECENTS_LIMIT) {
        break;
      }
    }
    return reordered;
  }
}
